ocrd/ocrd/workspace.py

```diff
@@ -461,64 +417,124 @@ def image_from_page(self, page, page_id,
-                     "AlternativeImage" if alternative_image else
-                     "image", page_id, orientation)
-            page_image = transpose_image(page_image, {
-                90: Image.ROTATE_90,
-                180: Image.ROTATE_180,
-                270: Image.ROTATE_270
-            }.get(orientation)) # no default
-            page_coords['features'] += ',rotated-%d' % orientation
-        if (orientation and
-            not 'rotated-%d' % orientation in feature_filter.split(',')):
-            # FIXME we should enforce consistency here (i.e. split into transposition
-            #       and minimal rotation)
-            if not (page_image.width == page_xywh['w'] and
-                    page_image.height == page_xywh['h']):
-                log.error('page "%s" image (%s; %dx%d) has not been transposed properly (%dx%d) during rotation',
+        # adjust the coord transformation to the steps applied on the image,
+        # and apply steps on the existing image in case it is missing there,
+        # but traverse all steps (crop/reflect/rotate) in a particular order:
+        # - existing image features take priority (in the order annotated),
+        # - next is cropping (if necessary but not already applied),
+        # - next is reflection (if necessary but not already applied),
+        # - next is rotation (if necessary but not already applied).
+        # This helps deal with arbitrary workflows (e.g. crop then deskew,
+        # or deskew then crop), regardless of where images are generated.
+        alternative_image_features = page_coords['features'].split(',')
+        for i, feature in enumerate(alternative_image_features +
+                                    (['cropped']
+                                     if (border and
+                                         not 'cropped' in page_coords['features'] and
+                                         not 'cropped' in feature_filter.split(','))
+                                     else []) +
+                                    (['rotated-%d' % orientation]
+                                     if (orientation and
+                                         not 'rotated-%d' % orientation in page_coords['features'] and
+                                         not 'rotated-%d' % orientation in feature_filter.split(','))
+                                     else []) +
+                                    (['deskewed']
+                                     if (skew and
+                                         not 'deskewed' in page_coords['features'] and
+                                         not 'deskewed' in feature_filter.split(','))
+                                     else []) +
+                                    # not a feature to be added, but merely as a fallback position
+                                    # to always enter loop at i == len(alternative_image_features)
+                                    ['_check']):
```

Comment:
That's one massive for-in-clause but I get the idea, though not this last part, even with the comment. You're adding a *pseudo-feature* `"_check_"` - why? How does that pertain to the loop index being equal to the number of features given?

Reply:
> How does that pertain to the loop index being equal to the number of features given?

…to the number of _already_ existing features.

Image size must be checked _after_ all existing features (from the AlternativeImage) have been reconstructed in the transform, but _before_ applying missing features (because operations in core are always consistent from there on, and I don't want to repeat the warning about the AlternativeImage multiple times).

But in case there are no missing features, I have to make sure to at least enter the loop body once more.

At this point, I know what you are going to say: why not just insert the pseudo-feature _before_ the added features and check for it directly instead of the position? – Consider it done!

Reply:
thx, I get it now. Yes, making the image size `_check` explicit is a good idea though.